doc/adr/0003-distributed-tracing.md

```diff
@@ -0,0 +1,158 @@
+# 3. NATS Service Latency Distributed Tracing Interoperability
+
+Date: 2020-05-21
+Author: @ripienaar
+
+## Status
+
+Proposed
+
+## Context
+
+The goal is to enable the NATS internal latencies to be exported to distributed tracing systems, here we see a small
+architecture using Traefik, a Go microservice and a NATS hosted service all being observed in Jaeger.
+
+![Jaeger](0003-jaeger-trace.png)
+
+The lowest 3 spans were created from a NATS latency Advisory.
+
+These traces can be ingested by many other commercial systems like Data Dog and Honeycomb where they can augment the
+existing operations tooling in use by our users. Additionally Grafana 7 supports Jaeger and Zipkin today.  
+
+Long term I think every server that handles a message should emit a unique trace so we can also get visibility into
+the internal flow of the NATS system and exactly which gateway connection has a delay - see our current HM issues - but
+ultimately I don't think we'll be doing that in the hot path of the server, though these traces are easy to handle async
+
+Meanwhile, this proposal will let us get very far with our current Latency Advisories.
+
+## Configuring an export
+
+Today there are no standards for the HTTP headers that communicate span context downstream - with Trace Context being
+an emerging w3c standard.
+
+I suggest we support the Jaeger and Zipkin systems as well as Trace Context for long term standardisation efforts.
+
+Supporting these would mean we have to interpret the headers that are received in the request to determine if we should
+publish a latency advisory rather than the static `50%` configuration we have today.
+
+Today we have:
+
+```
+exports: [
+    {
+        service: weather.service 
+        accounts: [WEB]
+        latency: {
+            sampling: 50% 
+            subject: weather.latency
+        }
+    }
+]
+```
+
+This enables sampling based `50%` of the service requests on this service.
+
+I propose we support additional sampling values `zipkin`, `jaeger`, `trace_context` which will configure the server to
+interpret the headers as below to determine if a request should be sampled.
+
+## Propagating headers
+
+The `io.nats.server.metric.v1.service_latency` advisory gets updated with additional `trace_format` and `headers` fields.
+
+`headers` would just propagate all the headers unmodified. We might later add support for a whitelist here to avoid
+leaking sensitive information.
+
+```json
+{
+  "type": "io.nats.server.metric.v1.service_latency",
+  "id": "YBxAhpUFfs1rPGo323WcmQ",
+  "timestamp": "2020-05-21T08:06:29.4981587Z",
+  "status": 200,
+  "trace_format": "jaeger",
+  "headers": {
```

Comment:
So we would move these from headers to a non-header section correct?

Reply:
What other headers would we need?

Reply:
all the headers the requestor sent in the request

Reply:
Do we need them or just the tracing based ones?

Reply:
I think there's value in all, but can go either way - I assumed all in case you didnt want to parse the whole lot and reencode them etc

Reply:
We have to anyway if you move them from headers to part of json body for latency metric.

Reply:
you could send the bytes only, but probably have to unpack them to decide to trace or not.

Is headers standardised to map[string]string or are we not forcing a format?

Reply:
We will need to agree on standard between clients and the servers, yes. Working on that now.

Reply:
To avoid unpacking header just to see if a particular value is present is why I've been advocating a flag to communicate that header contain values of interest to the server.